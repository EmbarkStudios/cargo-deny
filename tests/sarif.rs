use cargo_deny::{
    diag::{DiagnosticCode, Severity},
    sarif::SarifLog,
    sarif_collector::SarifCollector,
};

#[test]
fn test_sarif_structure() {
    let sarif = SarifLog::default();

    assert_eq!(
        sarif.schema,
        "https://json.schemastore.org/sarif-2.1.0.json"
    );
    assert_eq!(sarif.version, "2.1.0");
    assert!(sarif.runs.is_empty());
}

#[test]
fn test_sarif_collector() {
    let mut collector = SarifCollector::default();

    // Add a test diagnostic
    collector.add_diagnostic(
        DiagnosticCode::Advisory(cargo_deny::advisories::Code::Vulnerability),
        Severity::Error,
        "Test vulnerability found".to_string(),
        "test.rs".to_string(),
        42,
    );

    let sarif = collector.generate_sarif();

    // Verify structure
    assert_eq!(sarif.version, "2.1.0");
    assert_eq!(sarif.runs.len(), 1);

    let run = &sarif.runs[0];
    assert_eq!(run.tool.driver.name, "cargo-deny");
    assert_eq!(run.tool.driver.version, env!("CARGO_PKG_VERSION"));

    // Verify we have rules
    assert_eq!(run.tool.driver.rules.len(), 1);
    let rule = &run.tool.driver.rules[0];
    assert_eq!(rule.id, "Advisory(Vulnerability)");
    assert_eq!(rule.name, "Advisory(Vulnerability)");
    assert!(rule.short_description.text.contains("Security"));

    // Verify we have results
    assert_eq!(run.results.len(), 1);
    let result = &run.results[0];
    assert_eq!(result.rule_id, "Advisory(Vulnerability)");
    assert_eq!(result.message.text, "Test vulnerability found");
    assert_eq!(result.level, "error");

    // Verify location
    assert_eq!(result.locations.len(), 1);
    let location = &result.locations[0];
    assert_eq!(location.physical_location.artifact_location.uri, "test.rs");
    assert_eq!(location.physical_location.region.start_line, 42);
}

#[test]
fn test_sarif_multiple_diagnostics() {
    let mut collector = SarifCollector::default();

    // Add multiple diagnostics of different types
    collector.add_diagnostic(
        DiagnosticCode::Advisory(cargo_deny::advisories::Code::Vulnerability),
        Severity::Error,
        "Vulnerability 1".to_string(),
        "lib.rs".to_string(),
        10,
    );

    collector.add_diagnostic(
        DiagnosticCode::License(cargo_deny::licenses::Code::Unlicensed),
        Severity::Warning,
        "Missing license".to_string(),
        "main.rs".to_string(),
        20,
    );

    collector.add_diagnostic(
        DiagnosticCode::Bans(cargo_deny::bans::Code::Banned),
        Severity::Error,
        "Banned crate".to_string(),
        "Cargo.toml".to_string(),
        30,
    );

    let sarif = collector.generate_sarif();

    // Should have 3 different rules
    assert_eq!(sarif.runs[0].tool.driver.rules.len(), 3);

    // Should have 3 results
    assert_eq!(sarif.runs[0].results.len(), 3);

    // Verify each result has correct severity
    let results = &sarif.runs[0].results;
    assert!(
        results
            .iter()
            .any(|r| r.message.text == "Vulnerability 1" && r.level == "error")
    );
    assert!(
        results
            .iter()
            .any(|r| r.message.text == "Missing license" && r.level == "warning")
    );
    assert!(
        results
            .iter()
            .any(|r| r.message.text == "Banned crate" && r.level == "error")
    );
}

#[test]
fn test_sarif_json_serialization() {
    let mut collector = SarifCollector::default();

    collector.add_diagnostic(
        DiagnosticCode::Advisory(cargo_deny::advisories::Code::Notice),
        Severity::Note,
        "Notice".to_string(),
        "test.rs".to_string(),
        1,
    );

    let sarif = collector.generate_sarif();

    // Should be serializable to JSON
    let json = serde_json::to_string(&sarif).expect("Should serialize to JSON");
    assert!(json.contains("\"$schema\""));
    assert!(json.contains("\"version\":\"2.1.0\""));
    assert!(json.contains("cargo-deny"));

    // Should be deserializable back
    let parsed: SarifLog = serde_json::from_str(&json).expect("Should deserialize from JSON");
    assert_eq!(parsed.version, sarif.version);
    assert_eq!(parsed.runs.len(), sarif.runs.len());
}

#[test]
fn test_sarif_empty_diagnostics() {
    // Edge case: no diagnostics collected
    let collector = SarifCollector::default();
    let sarif = collector.generate_sarif();

    // Should still produce valid SARIF structure
    assert_eq!(
        sarif.schema,
        "https://json.schemastore.org/sarif-2.1.0.json"
    );
    assert_eq!(sarif.version, "2.1.0");
    assert_eq!(sarif.runs.len(), 1);

    let run = &sarif.runs[0];
    assert_eq!(run.tool.driver.name, "cargo-deny");
    assert_eq!(run.tool.driver.version, env!("CARGO_PKG_VERSION"));

    // Should have no rules and no results
    assert!(run.tool.driver.rules.is_empty());
    assert!(run.results.is_empty());

    // Should be serializable to valid JSON
    let json = serde_json::to_string(&sarif).expect("Should serialize to JSON");
    assert!(json.contains("\"$schema\""));
    assert!(json.contains("\"version\":\"2.1.0\""));
    assert!(json.contains("\"results\":[]"));
    assert!(json.contains("\"rules\":[]"));
}
