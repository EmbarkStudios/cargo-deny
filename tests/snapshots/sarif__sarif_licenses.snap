---
source: tests/sarif.rs
expression: s
---
{
  "$schema": "https://json.schemastore.org/sarif-2.1.0.json",
  "runs": [
    {
      "results": [
        {
          "level": "warning",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://{CWD}/examples/04_gnu_licenses/Cargo.toml"
                },
                "region": {
                  "byteLength": 7,
                  "byteOffset": 131,
                  "message": "a deprecated license identifier was used",
                  "snippet": "GPL-2.0",
                  "startLine": 4
                }
              }
            }
          ],
          "message": {
            "text": "error parsing SPDX license expression"
          },
          "partialFingerprints": {
            "cargo-deny/id": "l:parse-error",
            "cargo-deny/krate": "path+file://{CWD}/examples/04_gnu_licenses#gnu-licenses@0.1.0"
          },
          "ruleId": "l:parse-error"
        },
        {
          "level": "error",
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://{CWD}/examples/04_gnu_licenses/Cargo.toml"
                },
                "region": {
                  "byteLength": 51,
                  "byteOffset": 120,
                  "snippet": "Apache-2.0/GPL-2.0+ AND LGPL-3.0-only or gnu gpl v3",
                  "startLine": 4
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://{CWD}/examples/04_gnu_licenses/Cargo.toml"
                },
                "region": {
                  "byteLength": 10,
                  "byteOffset": 120,
                  "message": "rejected: license is not explicitly allowed",
                  "snippet": "Apache-2.0",
                  "startLine": 4
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://{CWD}/examples/04_gnu_licenses/Cargo.toml"
                },
                "region": {
                  "byteLength": 7,
                  "byteOffset": 131,
                  "message": "accepted: license is explicitly allowed",
                  "snippet": "GPL-2.0",
                  "startLine": 4
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://{CWD}/examples/04_gnu_licenses/Cargo.toml"
                },
                "region": {
                  "byteLength": 13,
                  "byteOffset": 144,
                  "message": "rejected: license is not explicitly allowed",
                  "snippet": "LGPL-3.0-only",
                  "startLine": 4
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "file://{CWD}/examples/04_gnu_licenses/Cargo.toml"
                },
                "region": {
                  "byteLength": 10,
                  "byteOffset": 161,
                  "message": "rejected: license is not explicitly allowed",
                  "snippet": "gnu gpl v3",
                  "startLine": 4
                }
              }
            }
          ],
          "message": {
            "text": "failed to satisfy license requirements"
          },
          "partialFingerprints": {
            "cargo-deny/id": "l:rejected",
            "cargo-deny/krate": "path+file://{CWD}/examples/04_gnu_licenses#gnu-licenses@0.1.0"
          },
          "ruleId": "l:rejected"
        }
      ],
      "tool": {
        "driver": {
          "name": "cargo-deny",
          "rules": [
            {
              "defaultConfiguration": {
                "level": "error"
              },
              "fullDescription": {
                "text": ""
              },
              "help": {
                "markdown": "[licenses check](https://embarkstudios.github.io/cargo-deny/checks/licenses/index.html)\n[rejected diagnostic](https://embarkstudios.github.io/cargo-deny/checks/licenses/diags.html#rejected)",
                "text": "For more information, see cargo-deny documentation"
              },
              "id": "l:rejected",
              "name": "l:rejected",
              "properties": {
                "precision": "high",
                "problem.severity": "error",
                "tags": [
                  "license",
                  "compliance"
                ]
              },
              "shortDescription": {
                "text": "A license was not explicitly accepted"
              }
            },
            {
              "defaultConfiguration": {
                "level": "warning"
              },
              "fullDescription": {
                "text": ""
              },
              "help": {
                "markdown": "[licenses check](https://embarkstudios.github.io/cargo-deny/checks/licenses/index.html)\n[parse-error diagnostic](https://embarkstudios.github.io/cargo-deny/checks/licenses/diags.html#parse-error)",
                "text": "For more information, see cargo-deny documentation"
              },
              "id": "l:parse-error",
              "name": "l:parse-error",
              "properties": {
                "precision": "high",
                "problem.severity": "warning",
                "tags": [
                  "license",
                  "compliance"
                ]
              },
              "shortDescription": {
                "text": "Failed to parse an SPDX expression"
              }
            }
          ],
          "semanticVersion": "0.18.4",
          "version": "0.18.4"
        }
      }
    }
  ],
  "version": "2.1.0"
}
